lib/Dialect/TTIR/Transforms/Passes.cpp

```diff
@@ -1407,4 +1456,82 @@ class TTIRLoadSystemDesc
   }
 };
 
+class TTIRGenericOpCBsRewriter : public OpRewritePattern<GenericOp> {
+public:
+  using OpRewritePattern<GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(GenericOp generic,
+                                PatternRewriter &rewriter) const final {
+    if (!generic.getOperandCbMapping().empty()) {
+      // Already inserted CBs, therefore skip.
+      return failure();
+    }
+
+    rewriter.setInsertionPointToStart(generic->getBlock());
+
+    SmallVector<Value> cbValues;
+    SmallVector<int64_t> operandCBMapping;
+
+    for (auto operand : generic->getOperands()) {
+      auto ty = mlir::cast<RankedTensorType>(operand.getType());
+
+      // Enforcing tiled layout as in kernel we always want to work with tiles.
+      auto desiredElementType = rewriter.getType<TileType>(ty.getElementType());
+      auto desiredLayout = rewriter.getAttr<LayoutAttr>(
+          ty, MemorySpace::DeviceL1, generic.getGrid(), desiredElementType);
+
+      auto operandTy = operand.getType();
+      auto operandLayout = mlir::cast<LayoutAttr>(
+          mlir::cast<RankedTensorType>(operandTy).getEncoding());
+
+      if (desiredLayout.getGrid() == operandLayout.getGrid()) {
+        // TODO(rpavlovic): should we check other layout features such as
+        // linear?
+        operandCBMapping.push_back(-1);
+        continue;
+      }
+
+      // Creating a CB for the operand. It takes the same type as the operand,
+      // but changes its grid. This may result in overly large CBs at the
+      // moment.
+      auto emptyOp = rewriter.create<tensor::EmptyOp>(
+          generic->getLoc(), ty.getShape(), ty.getElementType(), desiredLayout);
+      cbValues.push_back(emptyOp.getResult());
+      operandCBMapping.push_back(cbValues.size() - 1);
+    }
+
+    // TODO(rpavlovic): CBs could have constraint L1.
```

Comment:
Seems like they should inherit their associated input's constraint with L1 masked in and all other memory spaces masked out.

Reply:
Makes sense.